node/core/pvf/src/execute/queue.rs

```diff
@@ -155,6 +179,83 @@ impl Queue {
 			purge_dead(&self.metrics, &mut self.workers).await;
 		}
 	}
+
+	/// Returns a job index of a job that is supposed to be executed next, or `None` if the queue
+	/// is empty. If optional `idle_worker` is provided, it does its best to find a job that is
+	/// compatible with the worker's execution environment, but that is not guaranteed; the caller
+	/// must check if the job returned may or may not be executed with the given worker and must
+	/// kill the worker and spawn a new one to execute the job if needed.
+	/// The result may be used with `take_job()` to acquire the actual job.
+	fn next_job_index(&self, idle_worker: Option<Worker>) -> Option<usize> {
+		// New jobs are always pushed to the tail of the queue; the one at its head is always the eldest one.
+		// First check if we have a job that cannot wait any more even if we have to kill a worker to run it
+
+		let eldest = if let Some(eldest) = self.queue.get(0) { eldest } else { return None };
+
+		if eldest.waiting_since.elapsed() > MAX_KEEP_WAITING {
+			return Some(0)
+		}
```

Comment:
I noticed that we only take on waiting-too-long jobs when another job comes in, or one finishes. So if `MAX_KEEP_WAITING` is 60s, and another job comes in after say 120s (not sure how realistic this is), then the job would wait 120+s. By that point, the job might not even be relevant anymore (expired), but we always run it anyway. Is this a valid concern?

Reply:
I hope not, but maybe I'm missing something...
* If a job arrives and there's an idle worker with an incompatible environment, an idle worker is killed, and a new one is spawned and assigned the job (`else` branch in `try_assign_new_job()`)
* If a job arrives and there's a busy worker with a compatible environment, the job gets executed as soon as the worker finishes its current job (handled in `handle_job_finish()`)
* If a job arrives and there are busy works of incompatible types only
  * if the job becomes stale while the worker finishes its current job, it is executed by forcefully killing the worker and spawning a new one (handled with priority in `next_job_index()`, called by `try_assign_new_job()` called from `handle_job_finish()`)
  * if the job does not become stale while the worker finishes its current job, the job is still executed by forcefully killing the worker and spawning a new one, but it's taking another path (`handle_job_finish()` calls `try_assign_new_job()`, which calls `next_job_index()`, which tries to find a suitable worker for the eldest job, but, failing to find it, it returns the eldest job and instructs the caller to re-spawn the worker)

That is, I tried, by all means, to avoid a situation when a job just sits in the queue waiting to expire. I hope I succeeded, but I also see that the logic behind that came out rather complex and somewhat duplicated between `try_assign_next_job()` and `next_job_index()`. So it looks like I have to try to simplify it.

Reply:
My point is, if no job comes in or is killed for a while, then the check against `MAX_KEEP_WAITING` is not triggered. In that case, a job may expire in the meantime, but wait somewhat longer than the "max" time. So it is not really a "maximum". Sorry if I'm missing something, pretty tired today.

Reply:
Yes, it was a mess. I've finished reworking this part; please have a look at the new version, I hope it's much more straightforward now.

You're probably right about your concern; the "maximum" time may become overdue. But I don't think the job must be purged in that case anyway. Still, the comment describing `MAX_KEEP_WAITING` should be fixed.